### PR TITLE
Fix for read batch when as of is TimeStamp

### LINKED ISSUE #606: read(symbol, as_of) and read_batch([ReadRequest(symbol, as_of)]) behave differently for the same values of symbol and as_of
### Describe the bug

I chose "as_of" to be a datetime *just* after the timestamp of the version 0.

It seems read(symbol, as_of) and read_batch([ReadRequest(symbol, as_of)]) return different results (latter throws when former does not).

I would have expected both of them to return data (i.e. no exception).

### Steps/Code to Reproduce

```python

# lib = existing library with some data for symbol "X"

from arcticdb.version_store.library import ReadRequest
from datetime import datetime, timezone
vi = lib.list_versions("X")

# Let us assume this returns:
# X_v2: (date=2023-06-22 07:38:15.849084990+00:00),
# X_v1: (date=2023-06-22 05:41:31.169344280+00:00),
# X_v0: (date=2023-06-21 22:59:31.163315013+00:00)

as_of = datetime(2023, 6, 21, 22, 59, 32, tzinfo=timezone.utc) # Notice that this is just after the timestamp of v1
vi = lib.read("X", as_of=as_of) # this works fine.

vi = lib.read_batch([ReadRequest("X", as_of=as_of)]) 
# The above call throws an exception:
---------------------------------------------------------------------------
InternalException                         Traceback (most recent call last)
Input In [24], in <cell line: 1>()
----> 1 vi = lib.read_batch([ReadRequest("X", as_of=as_of)])

File /opt/bqplatform/lib/python3.9/site-packages/arcticdb/version_store/library.py:939, in Library.read_batch(self, symbols, query_builder)
    933     else:
    934         raise ArcticInvalidApiUsageException(
    935             f"Unsupported item in the symbols argument s=[{s}] type(s)=[{type(s)}]. Only [str] and"
    936             " [ReadRequest] are supported."
    937         )
--> 939 return self._nvs._batch_read_to_versioned_items(
    940     symbol_strings, as_ofs, date_ranges, columns, query_builder or query_builders
    941 )

File /opt/bqplatform/lib/python3.9/site-packages/arcticdb/version_store/_store.py:976, in NativeVersionStore._batch_read_to_versioned_items(self, symbols, as_ofs, date_ranges, columns, query_builder, kwargs)
    974 read_queries = self._get_read_queries(len(symbols), date_ranges, columns, query_builder)
    975 read_options = self._get_read_options(**kwargs)
--> 976 read_results = self.version_store.batch_read(symbols, version_queries, read_queries, read_options)
    977 versioned_items = []
    978 for i in range(len(read_results)):

InternalException: E_ASSERTION_FAILURE Version not found for stream


### Expected Results

Neither of the two methods should have thrown any exceptions.

### OS, Python Version and ArcticDB Version

Python: 3.9.12 | packaged by conda-forge | (main, Mar 24 2022, 23:22:55) 
[GCC 10.3.0]
OS: Linux-5.10.184-175.731.amzn2.x86_64-x86_64-with-glibc2.17
ArcticDB: 1.5.0

### Backend storage used

AWS S3

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by GemDot**
This PR fix a bug in the batch_get_versions_async function trying to get the version of specific timestamps. batch read method is affected by this issue.

close #606

### CODE REVIEW

File: python/tests/integration/arcticdb/test_arctic.py [hunk 1]
```diff
@@ -822,6 +822,49 @@ def test_write_batch_dedup(library_factory):
             assert data_key_version[s] == 0
 
 
+def test_read_batch_time_stamp(arctic_library):
+    """Should be able to write different size of batch of data."""
```
Comment (mehertz) [inline python/tests/integration/arcticdb/test_arctic.py:826]: Wrong comment?

File: python/tests/integration/arcticdb/test_arctic.py [hunk 2]
```diff
@@ -822,6 +822,49 @@ def test_write_batch_dedup(library_factory):
...
+        lib.write_batch(write_requests)
+        time.sleep(1)
+
+    data_non_batch = []
+    requests_batch = []
+    for sym_num in range(num_symbols):
+        versions = lib.list_versions(sym + str(sym_num))
+        for version_info in versions:
+            date_obj = versions[version_info].date
+            if date_obj.second == 59:
```
Comment (mehertz) [inline python/tests/integration/arcticdb/test_arctic.py:848]: https://docs.python.org/3/library/datetime.html#examples-of-usage-timedelta

848-858 can be removed I think.

File: python/tests/integration/arcticdb/test_arctic.py [hunk 3]
```diff
@@ -822,6 +822,49 @@ def test_write_batch_dedup(library_factory):
...
+                date_obj.second + 1,
+                tzinfo=timezone.utc,
+            )
+            data_non_batch.append(lib.read(sym + str(sym_num), as_of=as_of))
+            requests_batch.append(ReadRequest(sym + str(sym_num), as_of=as_of))
+
+    data_batch = lib.read_batch(requests_batch)
+    list_data_to_compate = list(zip(data_non_batch, data_batch))
+    for d1, d2 in list_data_to_compate:
+        assert_frame_equal(d1.data, d2.data)
```
Comment (mehertz) [inline python/tests/integration/arcticdb/test_arctic.py:865]: Can we also check that this is correct, not that it's the same behaviour as `read`.

**Review by mehertz — DISMISSED** [review]

File: python/tests/integration/arcticdb/test_arctic.py (hunk 3 above)
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py:868]: done the extra check

File: python/tests/integration/arcticdb/test_arctic.py (hunk 2 above)
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py]: removed

File: python/tests/integration/arcticdb/test_arctic.py (hunk 1 above)
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py]: amended

**Review by mehertz — APPROVED** [review]

File: python/tests/integration/arcticdb/test_arctic.py [hunk 4]
```diff
@@ -822,6 +822,52 @@ def test_write_batch_dedup(library_factory):
             assert data_key_version[s] == 0
 
 
+def test_read_batch_time_stamp(arctic_library):
```
Comment (poodlewars) [inline python/tests/integration/arcticdb/test_arctic.py:825]: We should stop adding more tests to `test_arctic.py` it's getting completely unwieldy

File: python/tests/integration/arcticdb/test_arctic.py [hunk 5]
```diff
@@ -822,6 +822,52 @@ def test_write_batch_dedup(library_factory):
...
+        time.sleep(1)
+
+    data_non_batch = []
+    requests_batch = []
+    original_dataframes = []
+    for sym_num in range(num_symbols):
+        versions = lib.list_versions(sym + str(sym_num))
+        for version_info in versions:
+            date_obj = versions[version_info].date
+            date_obj += +timedelta(seconds=1)
```
Comment (poodlewars) [inline python/tests/integration/arcticdb/test_arctic.py:849]: random extra `+`?

File: cpp/arcticdb/version/version_map_batch_methods.hpp [hunk 6]
```diff
@@ -192,8 +192,14 @@ inline std::optional<AtomKey> get_key_for_version_query(
         [&version_map_entry] (const pipelines::SpecificVersionQuery& specific_version) {
             return find_index_key_for_version_id(specific_version.version_id_, version_map_entry);
         },
-        [&version_map_entry] (const pipelines::TimestampVersionQuery& timestamp_version) {
-        return find_index_key_for_version_timestamp(timestamp_version.timestamp_, version_map_entry);
+        [&version_map_entry] (const pipelines::TimestampVersionQuery& timestamp_version) -> std::optional<AtomKey> {
+            auto version_key = get_version_key_from_time_for_versions(timestamp_version.timestamp_, version_map_entry->get_indexes(false));
+            if(version_key.has_value()){
+                auto version_id = version_key.value().version_id();
```
Comment (jjerphan) [inline cpp/arcticdb/version/version_map_batch_methods.hpp:198]: Could you please type `version_key` and `version_id` explicitly? :pray:

File: python/tests/integration/arcticdb/test_arctic.py [hunk 7]
```diff
@@ -822,6 +822,52 @@ def test_write_batch_dedup(library_factory):
...
+                tzinfo=timezone.utc,
+            )
+            data_non_batch.append(lib.read(sym + str(sym_num), as_of=as_of))
+            requests_batch.append(ReadRequest(sym + str(sym_num), as_of=as_of))
+            v_num = version_info.version
+            original_dataframes.append(pd.DataFrame({"col": [sym_num + v_num, sym_num * v_num, sym_num - v_num]}))
+
+    data_batch = lib.read_batch(requests_batch)
+    list_data_to_compate = list(zip(data_non_batch, data_batch, original_dataframes))
+    for d1, d2, d3 in list_data_to_compate:
```
Comment (jjerphan) [inline python/tests/integration/arcticdb/test_arctic.py:866]:
```suggestion
    for d1, d2, d3 in zip(data_non_batch, data_batch, original_dataframes):
```

File: python/tests/integration/arcticdb/test_arctic.py (hunk 4 above)
Comment (jjerphan) [reply python/tests/integration/arcticdb/test_arctic.py:825]: I agree with you. Yet I think we must test this behavior.

How should we proceed?

File: python/tests/integration/arcticdb/test_arctic.py (hunk 3 above)
Comment (poodlewars) [reply python/tests/integration/arcticdb/test_arctic.py:867]: Then what is the point of leaving the "is the same as `read`" check?

**Comment by poodlewars** [comment]
It would have been good to explain the cause of the problem and how this fixes it in the PR description. AFAICT the issue was that 

```
inline std::optional<AtomKey> find_index_key_for_version_timestamp(timestamp as_of, const std::shared_ptr<VersionMapEntry>& entry, bool included_deleted=true) {
    auto key = std::find_if(std::begin(entry->keys_), std::end(entry->keys_), [as_of] (const auto& key) {
        return is_index_key_type(key.type()) && key.creation_ts() == as_of;
    });
    if(key == std::end(entry->keys_))
        return std::nullopt;

    return included_deleted || !entry->is_tombstoned(*key) ? std::make_optional(*key) : std::nullopt;
}
```

was checking for exact equality with the provided as_of?

File: python/tests/integration/arcticdb/test_arctic.py [hunk 8]
```diff
@@ -822,6 +822,52 @@ def test_write_batch_dedup(library_factory):
...
+    for v_num in range(num_versions):
+        write_requests = []
+        for sym_num in range(num_symbols):
+            write_requests.append(
+                WritePayload(
+                    sym + str(sym_num), pd.DataFrame({"col": [sym_num + v_num, sym_num * v_num, sym_num - v_num]})
+                )
+            )
+        lib.write_batch(write_requests)
+        time.sleep(1)
```
[the remark this answers is not in the document]
Comment (mehertz) [reply python/tests/integration/arcticdb/test_arctic.py:840]: I can see we need some minimum delay, but 1s is quite a lot.

File: python/tests/integration/arcticdb/test_arctic.py (hunk 3 above)
Comment (mehertz) [reply python/tests/integration/arcticdb/test_arctic.py:867]: yeh that's fair

File: python/tests/integration/arcticdb/test_arctic.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexowens90) [reply python/tests/integration/arcticdb/test_arctic.py:840]: We don't need a sleep at all if you just record the time after each call to `write_batch` and then read as_of those times

File: cpp/arcticdb/version/version_map_batch_methods.hpp [hunk 9]
```diff
@@ -192,8 +192,14 @@ inline std::optional<AtomKey> get_key_for_version_query(
         [&version_map_entry] (const pipelines::SpecificVersionQuery& specific_version) {
             return find_index_key_for_version_id(specific_version.version_id_, version_map_entry);
         },
-        [&version_map_entry] (const pipelines::TimestampVersionQuery& timestamp_version) {
-        return find_index_key_for_version_timestamp(timestamp_version.timestamp_, version_map_entry);
```
[the remark this answers is not in the document]
Comment (GemDot) [reply cpp/arcticdb/version/version_map_batch_methods.hpp:196]: This is dead code. So, we should remove this logic. Good catch. Note that we are reusing another function from the non-batch method, in order to achieve the desired behaviour

File: cpp/arcticdb/version/version_map_batch_methods.hpp (hunk 6 above)
Comment (GemDot) [reply cpp/arcticdb/version/version_map_batch_methods.hpp:198]: we prefer auto  as a style in the code base. I will politely dismiss your comment here

File: python/tests/integration/arcticdb/test_arctic.py (hunk 4 above)
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py:825]: done in https://github.com/man-group/ArcticDB/pull/651/files#diff-cbe6d7a5a240c5c6068d4319fe88bbfbee9434ff8e4cbc84f6f68b26535005fa

File: python/tests/integration/arcticdb/test_arctic.py (hunk 5 above)
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py:849]: done in https://github.com/man-group/ArcticDB/pull/651/files#diff-cbe6d7a5a240c5c6068d4319fe88bbfbee9434ff8e4cbc84f6f68b26535005fa

File: python/tests/integration/arcticdb/test_arctic.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py:840]: done in https://github.com/man-group/ArcticDB/pull/651/files#diff-cbe6d7a5a240c5c6068d4319fe88bbfbee9434ff8e4cbc84f6f68b26535005fa

File: python/tests/integration/arcticdb/test_arctic.py [hunk 10]
```diff
@@ -822,6 +822,52 @@ def test_write_batch_dedup(library_factory):
...
 
+def test_read_batch_time_stamp(arctic_library):
+    """Should be able to read data in batch mode using a timestamp."""
+    lib = arctic_library
+    sym = "sym_"
+    num_versions = 3
+    num_symbols = 3
+    for v_num in range(num_versions):
+        write_requests = []
+        for sym_num in range(num_symbols):
```
[the remark this answers is not in the document]
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py:833]: done in https://github.com/man-group/ArcticDB/pull/651/files#diff-cbe6d7a5a240c5c6068d4319fe88bbfbee9434ff8e4cbc84f6f68b26535005fa

File: python/tests/integration/arcticdb/test_arctic.py (hunk 7 above)
Comment (GemDot) [reply python/tests/integration/arcticdb/test_arctic.py:866]: done in https://github.com/man-group/ArcticDB/pull/651/files#diff-cbe6d7a5a240c5c6068d4319fe88bbfbee9434ff8e4cbc84f6f68b26535005fa

**Pull request merged**